l1-contracts/contracts/bridge/L1ERC20Bridge.sol

```diff
@@ -167,39 +390,100 @@ contract L1ERC20Bridge is IL1Bridge, IL1BridgeLegacy, ReentrancyGuard {
         }
-        l2TxHash = zkSync.requestL2Transaction{value: msg.value}(
-            l2Bridge,
-            0, // L2 msg.value
+
+        l2TxHash = _depositSendTx(
+            _chainId,
+            _mintValue,
             l2TxCalldata,
             _l2TxGasLimit,
             _l2TxGasPerPubdataByte,
-            new bytes[](0),
             refundRecipient
         );
 
         // Save the deposited amount to claim funds on L1 if the deposit failed on L2
-        depositAmount[msg.sender][_l1Token][l2TxHash] = amount;
+        bytes32 txDataHash = keccak256(abi.encode(msg.sender, _l1Token, _amount));
+        deposited[_chainId][txDataHash][l2TxHash] = true;
+
+        emit DepositInitiatedSharedBridge(_chainId, txDataHash, msg.sender, _l2Receiver, _l1Token, _amount);
+        if (_chainId == ERA_CHAIN_ID) {
+            emit DepositInitiated(l2TxHash, msg.sender, _l2Receiver, _l1Token, _amount);
+        }
+    }
+
+    // to avoid stack too deep error
+    function _depositSendTx(
+        uint256 _chainId,
+        uint256 _mintValue,
+        bytes memory _l2TxCalldata,
+        uint256 _l2TxGasLimit,
+        uint256 _l2TxGasPerPubdataByte,
+        address _refundRecipient
+    ) internal returns (bytes32 l2TxHash) {
+        // note msg.value is 0 for not eth base tokens.
+
+        L2TransactionRequestDirect memory request = L2TransactionRequestDirect({
+            chainId: _chainId,
+            l2Contract: l2BridgeAddress[_chainId],
+            mintValue: _mintValue, // l2 gas + l2 msg.Value the bridgehub will withdraw the mintValue from the base token bridge for gas
+            l2Value: 0, // L2 msg.value, this contract doesn't support base token deposits or wrapping functionality, for direct deposits use bridgehub 
+            l2Calldata: _l2TxCalldata,
+            l2GasLimit: _l2TxGasLimit,
+            l2GasPerPubdataByteLimit: _l2TxGasPerPubdataByte,
+            factoryDeps: new bytes[](0),
+            refundRecipient: _refundRecipient
+        });
 
-        emit DepositInitiated(l2TxHash, msg.sender, _l2Receiver, _l1Token, amount);
+        l2TxHash = bridgehub.requestL2Transaction{value: msg.value}(request);
+    }
+
+    /// used by bridgehub to aquire mintValue. If l2Tx fails refunds are sent to refundrecipient on L2
+    function bridgehubDepositBaseToken(
+        uint256 _chainId, // we will need this for Marcin's idea: tracking chain's balances until hyperbridging
+        address _prevMsgSender,
+        address _l1Token,
+        uint256 _amount
+    ) public payable onlyBridgehub {
+        require(msg.value == 0, "L1EB: msg.value > 0 base deposit"); // this bridge does not hold eth, the weth bridge does
+        require(_amount != 0, "4T"); // empty deposit amount
+        if (_prevMsgSender != address(this)){ // the bridge might be calling itself, in which case the funds are already in the contract. This only happens in testing, as we will not support the ERC20 contract as a base token bridge
```

Comment:
If it happens only in testing, maybe it is possible to remove it with the preprocessor?

Reply:
Is it a problem to keep it in though? This is valid either way. Adding preprocessors would only make it more complicated, no? Wdyt?

Reply:
We typically try to avoid things that are only needed for testing from getting into production (this is why we have multiple bootloaders for different purposes instead of 1 bootloader with config params).

In this particular place, you've already used the preprocessed (for if governance is EOA) and so adding a preprocessor shouldnt add any complexity

Reply:
Ok, done